Feature/kan 2

denmark-dangnagui opened the pull request:
카카오 OAuth 구현해봤습니다.

확인 한 번 해주시면 감사합니다~~

- close #2

denmark-dangnagui committed:
KAN-2 테스트

denmark-dangnagui committed:
카카오 OAuth 구현

somefood commented:
구현해주셔서 감사합니다.
한 번 확인하고 궁금한 점에 대해 여쭤보겠습니다~!

somefood commented on README.md:
```diff
@@ -1,2 +1,4 @@
 # sulsul-back
 sulsul-back
+
+홀롤로롤
```
지워주시기 바랍니다~!

denmark-dangnagui replied:
넵! PR 테스트 용도로 해뒀던 거라서 지우고 반영하도록 하겠습니다~!

somefood commented on src/main/resources/db/migration/V20240114_2205__create_member_table.sql:
```diff
@@ -1,10 +1,12 @@
+    access_token    VARCHAR(255),
+    created_at      DATETIME,
+    updated_at      DATETIME,
+    CONSTRAINT idx_username UNIQUE (nickname)
```
flyway는 DDL의 이력을 관리하는 것이니 컬럼의 변경이 일어난다면 새로운 파일로 DDL을 만들어 주시기 바랍니다!
아직은 초창기니까 괜찮은데, 파일을 수정하는 경우 관리의 의미가 없어질 수 있거든요

아래 링크 참고해서 보시면 좋을거 같습니다 ㅎㅎ
https://hudi.blog/dallog-flyway/

denmark-dangnagui replied:
그렇군요,,, 일단 좀 더 저도 공부를 해봐야될 것 같습니다! 우선 링크 참조해서 알고 있겠습니다! 감사합니다 :)

somefood commented on src/main/resources/application.yml:
```diff
@@ -1,9 +1,19 @@
+      - jwt
+  oauth:
+    kakao:
+      secret: YmkpW6daM1UdRVPcLUx979MAaFNiIVt5
+      client-id: e0df3d8e0617f4bba3f0c3c909eef6ea
```
해당 키들은 public 한 곳에 노출되면 보안이 취약해지니 지워주시기 바랍니다.
또한, 기존 애플리케이션도 다시 만들어 키를 갱신해주는게 좋을거 같습니다~! :)

denmark-dangnagui replied:
넵! 알겠습니다! 이것까지 올리고 말았네요 ㅠ
흠, 근데 애플리케이션을 새로 만드는건 큰 문제는 되지 않지만, 이미 public한 곳에 노출이 되었으니, 새롭게 만드는게 보안상 좋으니 갱신하면 좋겠다는 말씀이실까요??

somefood replied:
넵, 보안키들은 항상 악용될 소지가 다분하기에 테스트용도라 할지라도 노출 안 시키는게 좋다고 생각합니다~!

somefood commented on src/main/resources/application-jwt.yml:
```diff
@@ -0,0 +1,15 @@
+      on-profile: local
+
+jwt:
+  secretKey: testSecretKey20230327testSecretKey20230327testSecretKey202303271221122121212122121212112212121221212112211212121212121221
```
JWT 시크릿키가 노출되면 외부에서 위변조가 가능하니 숨겨주시면 감사하겠습니다! ㅎㅎ

denmark-dangnagui replied:
잉것도 숨겨놓도록 하겠습니다 ㅎㅎ,,🥲

somefood commented on src/main/java/main/sulsul/authentication/infra/kakao/KakaoInfoResponse.java:
```diff
@@ -0,0 +1,45 @@
+import main.sulsul.authentication.domain.oauth.OAuthProvider;
+
+@Getter
+@JsonIgnoreProperties(ignoreUnknown = true)
```
이 어노테이션과 속성의 용도가 어떤 건지 궁금합니다!

denmark-dangnagui replied:
저도 한번 찾아보고 알게되었는데요! 흠 쉽게말하면 JSON 데이터를 읽어 객체에 매핑해주는 과정에서 객체에는 없어서 알 수 없는 property가 JSON 데이터에 있어도 에러를 내지 않게 하는 어노테이션입니다! 

예를 들어 클라이언트 측에서 서버로 JSON 형식으로 넘길 때 객체에는 없는 key:value를 보내더라도 에러가 발생하지 않게 하도록 하는 것입니다!  

우선 api 개발할 때 테스트 목적이 짙어 해당 어노테이션도 한번 사용해봤습니다! 😁

somefood commented on src/main/java/main/sulsul/authentication/domain/oauth/RequestOAuthInfoService.java:
```diff
@@ -0,0 +1,25 @@
+public class RequestOAuthInfoService {
+    private final Map<OAuthProvider, OAuthApiClient> clients;
+
+    public RequestOAuthInfoService(List<OAuthApiClient> clients) {
+        this.clients = clients.stream().collect(
+                Collectors.toUnmodifiableMap(OAuthApiClient::oAuthProvider, Function.identity())
+        );
+    }
```
생성자로 동일 타입들 리스트로 받아서 맵으로 등록하는 방식 잘 봤습니다.
이렇게 하면 다형성을 잘 이용할 수 있겠네요~!

somefood commented on src/main/java/main/sulsul/authentication/domain/oauth/RequestOAuthInfoService.java:
```diff
@@ -0,0 +1,25 @@
+
+    public RequestOAuthInfoService(List<OAuthApiClient> clients) {
+        this.clients = clients.stream().collect(
+                Collectors.toUnmodifiableMap(OAuthApiClient::oAuthProvider, Function.identity())
```
추가로 Function.identity() 메서드 설명해주시면 감사하겠습니다 ㅎㅎ

somefood commented on src/main/java/main/sulsul/authentication/domain/oauth/RequestOAuthInfoService.java:
```diff
@@ -0,0 +1,25 @@
+        );
+    }
+
+    public OAuthInfoResponse request(OAuthLoginParams params) {
+        OAuthApiClient client = clients.get(params.oAuthProvider());
+        String accessToken = client.requestAccessToken(params);
+        return client.requestOauthInfo(accessToken);
```
다형성을 잘 두르시는군요. 참고하겠습니다~!

denmark-dangnagui replied:
여기저기서 많이 검색해보면서 저런식으로 코드를 짜게 되었네요 ㅎㅎ,, 
앞으로 더 공부해서 저런식으로 코드를 짜고 더욱 좋은 방법을 계속해서 추구하겠습니다!

somefood commented on build.gradle:
```diff
@@ -23,10 +23,15 @@ repositories {
+    implementation 'io.jsonwebtoken:jjwt-jackson:0.11.5'
+    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+
+    implementation 'org.springframework.boot:spring-boot-starter-oauth2-client'
```
현재 코드 보았을땐 oAuth 라이브러리는 필요없어 보이는데,
어디서 사용되고 있는지 알 수 있을까요~?

denmark-dangnagui replied:
오 맞습니다 제가 맨처음엔 해당 라이브러리를 사용해서 테스트를 진행했었는데, 구현을 시도해보니 아직은 어려워서 좀 더 공부를 해본 다음에 구현하는게 성장에 도움이 될 것 같아 프로젝트에 영향도면에서 일단은 소셜로그인만 구축해보려고 했습니다! 그러다 보니 현재 코드를 사용하게 되었는데 지우도록 하겠습니다! 

 추후 시큐리티랑 같이 적용해볼 생각이라, 해당 라이브러리는 필요시에 적용하는 것으로 하겠습니다! 체크해주셔서 감사합니다! 🫡

somefood commented on src/main/java/main/sulsul/authentication/controller/MemberController.java:
```diff
@@ -0,0 +1,39 @@
+        return ResponseEntity.ok(memberRepository.findAll());
+    }
+
+    @GetMapping("/{accessToken}")
+    public ResponseEntity<Member> findByAccessToken(@PathVariable(value = "accessToken") String accessToken) {
+        Long memberId = authTokensGenerator.extractMemberId(accessToken);
+        return ResponseEntity.ok(memberRepository.findById(memberId).get());
+    }
```
보통 JWT의 토큰은 Request Header의 Authorization Bearer ${accessToken} 형태로 넣고 있으니, 
컨트롤러에서 액세스 토큰을 인자로 받는 것이 아닌 위의 형태로 바꿔서 사용할 수 있게 바꿔주는게 좋을거 같습니다.
이렇게 하시면서 필터나 인터셉터를 하나 도입해서 해당 토큰이 유효한지에 대한 검증을 하는 작업도 해주시면 되겠습니다.

추가적으로 각 컨트롤러에서 JWT 토큰에 있는 ID 값을 어떻게 꺼내 쓸 수 있는지 대한 설명이 필요할거 같습니다! (필수)

denmark-dangnagui replied:
오 넵! 일단 간단한 테스트 용도로만 사용을 해본거라 저런식으로 경로 인자값으로 들어가게 되었습니다 ㅠ,, 

해당 api는 수정하도록 하겠습니다! 

그리고 ID 값을 꺼내는 것은 해당 넘어오는 토큰에 있는 클레임을 분석해서 결과적으로 토큰에 있는 이메일 정보를 통해서 해당 ID값을 가져오도록 되어있습니다!! 아마 계속해서 이러한 방식을 사용할지는 좀 더 고민해보도록 하겠습니다!

somefood commented on src/main/java/main/sulsul/authentication/domain/AuthTokens.java:
```diff
@@ -0,0 +1,22 @@
+    private String grantType;
+    private Long expiresIn;
+
+    public static AuthTokens of(String accessToken, String refreshToken, String grantType, Long expiresIn) {
+        return new AuthTokens(accessToken, refreshToken, grantType, expiresIn);
+    }
```
스태틱 메서드를 활용해서 인스턴스를 제공할거면
AllArgsConstrutor의 액세스 레벨을 private로 바꿔도 좋을거 같습니다~

somefood commented on src/main/java/main/sulsul/authentication/domain/Member.java:
```diff
@@ -0,0 +1,33 @@
+    @Enumerated(EnumType.STRING)
+    @Column(name = "o_auth_provider", length = 20)
+    private OAuthProvider oAuthProvider;
+
+
```
추후 스프링 시큐리티 기능을 활용하기 위해 Role 개념을 도입하고자 합니다.
Role enum과 함께 DDL도 한 번 추가해주세요!
그리고 카카오로 회원가입하는 경우엔 Role은 USER로 해주시기 바랍니다!

denmark-dangnagui replied:
넵! 우선 전체적으로 수정은 해보고 있습니다! 
일단 시큐리티 기능을 제하고 작업은 해보고 있는데 말씀하신대로 추후에는 시큐리티 기능을 추가할 것이니, 오늘 ddl 만들고 entity도 수정할 것이라, 내일 간이 회의로 한번 봐주시면 감사할 것 같습니다! 😄

somefood commented:
추가적으로 패키지 구조도 한 번 손 볼까 합니다.
authentication이란 패키지 대신엔 member랑 oauth 패키지로 분리해서 바꿔보는게 어떨까요~?
사용자 로그인이긴 하지만 둘의 도메인은 다르니 나눠보는것도 괜찮을거 같습니다.
또한, controller, service 대신에 각각 ui, application으로 한 번 바꿔보는것도 어떨지요?
좀 더 포괄적인 이름으로 담아볼까 합니다

아래는 다른 사플의 구조인데, 예시 첨부드립니다!
![image](https://github.com/team-sulsul/sulsul-back/assets/24751937/2d1a8921-9a73-446e-a1f2-6bee89a46537)

![image](https://github.com/team-sulsul/sulsul-back/assets/24751937/5bfc8009-873a-46da-887c-86d22c4ba30e)

somefood commented:
현재 develop에 타임리프 적용까지 올려두었는데, 리베이스해서 작업 부탁드릴게요~~!
패키지라던가, 파일 구조들 보고 잘 배치해주심 됩니다~

denmark-dangnagui committed:
카카오 OAuth 구현

- refresh 토큰 값 저장 로직 추가
- ROLE 멤버 엔티티에 추가
- 토큰 유효성 검사 추가
- 토큰 값 경로 추가 아닌, 바디값으로 리퀘스트받게 변경

denmark-dangnagui committed:
DDD에 최대한 맞춰서 패키지 구조 변경 및 pr에 코멘트에 있었던 것들 반영하려고 하였습니다~!

somefood commented on src/main/java/main/sulsul/member/domain/Role.java:
```diff
@@ -1,7 +1,16 @@
+
+@Getter
 public enum Role {
+    USER("카카오유저"),
```
멤버의 Role에 대한 메시지를 적으신거 같은데, USER의 Role은 범용적이라 생각하기에 카카오 유저대신 더 포괄적으로 담아주시는게 좋을거 같아요~!

+ ADMIN Role도 추가해주시면 좋겠습니다!

denmark-dangnagui replied:
현재 시큐리티에 대해서 공부를 하고 있습니다! 많이 부족하지만, 이젠 ROLE 에 ADMIN 이 필요한 이유도 알게되어서 추가 적용하여 다시 PR 올리도록 하겠습니다~! 감사합니😁

somefood commented on src/main/java/main/sulsul/authentication/controller/MemberController.java:
```diff
@@ -11,29 +12,26 @@
 @RestController
 @RequiredArgsConstructor
-@RequestMapping("/api/members")
+@RequestMapping("/api/member")
```
복수형태에서 단수 형태로 바꾸신 이유가 궁금합니다.

denmark-dangnagui replied:
이유가 크지는 않습니다! 중복된 컨트롤러 경로를 갖게 되어 아마 그랬던 것 같은데, 추가적으로 적용하면서 좀 더 의미있는 경로명을 가지도록 수정할게요!

somefood commented on src/main/java/main/sulsul/authentication/controller/MemberController.java:
```diff
@@ -11,29 +12,26 @@
+    @GetMapping("/all")
     public ResponseEntity<List<Member>> findAll() {
         return ResponseEntity.ok(memberRepository.findAll());
     }
```
해당 컨트롤러는 어느 시점에 사용되는건지 알 수 있을까요?

denmark-dangnagui replied:
우선 필요한 것이라기 보단 계속해서 테스트를 하고 있다보니 아직 남아있던 컨트롤러 같아요! 
추후 필요없어질 시기에는 해당 컨트롤러는 없애도록 하겠습니다~!

somefood commented on src/main/java/main/sulsul/authentication/domain/AuthTokensGenerator.java:
```diff
@@ -24,6 +27,13 @@ public AuthTokens generate(Long memberId) {
+        newMember.ifPresent(member -> {
+            member.setRefreshToken(refreshToken);
+            memberRepository.save(member); // 업데이트된 Member를 저장
+            System.out.println("Updated Member: " + member);
```
print 출력보단 slf4j 활용해서 찍히도록 변경해주세요!

denmark-dangnagui replied:
아직도 print문이 편한게 한스럽네요 slf4j를 사용하도록 습관들이도록 하겠음다!

somefood commented on src/main/java/main/sulsul/authentication/infra/JwtTokenProvider.java:
```diff
@@ -43,5 +46,40 @@ private Claims parseClaims(String accessToken) {
+
+    public static boolean isAccessTokenValid(String accessToken) {
+        try {
+            System.out.println("토큰이 정상적입니다.");
```
print 출력보단 slf4j 활용해서 찍히도록 변경해주세요!

denmark-dangnagui replied:
넵!

somefood commented on src/main/java/main/sulsul/authentication/infra/JwtTokenProvider.java:
```diff
@@ -43,5 +46,40 @@ private Claims parseClaims(String accessToken) {
+            return isTokenExpired(accessToken);
+        } catch (ExpiredJwtException e) {
+            // AccessToken이 만료된 경우
+            System.out.println("AccessToken이 만료되었습니다.");
```
print 출력보단 slf4j 활용해서 찍히도록 변경해주세요!

somefood commented on src/main/java/main/sulsul/authentication/infra/JwtTokenProvider.java:
```diff
@@ -43,5 +46,40 @@ private Claims parseClaims(String accessToken) {
+            return false;
+        } catch (MalformedJwtException | SignatureException e) {
+            // 유효하지 않은 형식의 토큰 또는 서명 오류가 발생한 경우
+            System.out.println("유효하지 않은 AccessToken입니다.");
```
print 출력보단 slf4j 활용해서 찍히도록 변경해주세요!

somefood commented on src/main/java/main/sulsul/authentication/infra/JwtTokenProvider.java:
```diff
@@ -29,6 +31,7 @@ public String generate(String subject, Date expiredAt) {
         Claims claims = parseClaims(accessToken);
+        isAccessTokenValid(accessToken);
         return claims.getSubject();
     }
```
Subject를 추출할때 isAccessTokenValid 메서드의 반환 타입이 boolean 인데, 반환값을 활용하는거 같진 않아 보입니다.
어떻게 사용되는지 궁금합니다~!

denmark-dangnagui replied:
말씀해주신걸 보고 코드를 보니 굳이 boolean타입으로 갈 필요는 없겠군요! 
그 이유는 해당 메서드안에서 예외처리가 됩니다! 
반환 타입을 변경해두도록 하겠습니다!

somefood commented on src/main/resources/application.yml:
```diff
@@ -6,8 +6,8 @@ spring:
     kakao:
-      secret: YmkpW6daM1UdRVPcLUx979MAaFNiIVt5
-      client-id: e0df3d8e0617f4bba3f0c3c909eef6ea
+      secret:
+      client-id:
```
OAuth 정보들 공유와 더불어 어떻게 값 삽입하면 되는지 공유 부탁드립니다~!

denmark-dangnagui replied:
일전에 석주님이 말씀해주셨던 것 처럼 해당 값들을 깃헙에서 관리자들만 공유하던게 있었던 것 같은데,, 한번 제가 말씀해주신 걸 토대로 적용해서 말씀드리도록 하겠습니다!

somefood commented on src/main/resources/application-jwt.yml:
```diff
@@ -4,7 +4,7 @@ spring:
 
 jwt:
-  secretKey: testSecretKey20230327testSecretKey20230327testSecretKey202303271221122121212122121212112212121221212112211212121212121221
+  secretKey:
```
시크릿키 공유와 사용 방법 공유 부탁드립니다~!

denmark-dangnagui replied:
위에서 말씀드렸다시피 한번 찾아보고 말씀드리겠습니다!

denmark-dangnagui committed:
불필요한 import 제거 및 패키지 정리

denmark-dangnagui committed:
Member 테이블에 Oauth용 컬럼 추가
